### PR TITLE
[CALCITE-4066] SqlTypeUtil#convertTypeToSpec cover Array/Multiset/Row types.

### PR DESCRIPTION
**Pull request by Aaaaaaron**
Add array/multiset/row type support.

### CODE REVIEW

File: core/src/test/java/org/apache/calcite/sql/type/SqlTypeUtilTest.java [hunk 1]
```diff
@@ -117,6 +131,57 @@
     SqlTypeCoercionRule.THREAD_PROVIDERS.set(defaultRules);
   }
 
+  @Test void testConvertTypeToSpec() {
+    final SqlTypeFactoryImpl factory =
+        new SqlTypeFactoryImpl(RelDataTypeSystem.DEFAULT);
+    final RelDataType bigint = factory.createSqlType(SqlTypeName.BIGINT);
```
Comment (danny0405) [inline core/src/test/java/org/apache/calcite/sql/type/SqlTypeUtilTest.java:137]: Try to reuse the type defined in `f`,if there is no what you need, create one there.

Comment (Aaaaaaron) [reply core/src/test/java/org/apache/calcite/sql/type/SqlTypeUtilTest.java:137]: Thanks for your advice.

File: core/src/test/java/org/apache/calcite/sql/type/SqlTypeUtilTest.java [hunk 2]
```diff
@@ -117,6 +126,37 @@
...
+        .stream()
+        .map(SqlIdentifier::getSimple)
+        .collect(Collectors.toList());
+    List<String> fieldTypeNames = rowSpec.getFieldTypes()
+        .stream()
+        .map(f -> f.getTypeName().getSimple())
+        .collect(Collectors.toList());
+    assertThat(rowSpec.getTypeName().getSimple(), is("ROW"));
+    assertThat(fieldNames, is(Lists.newArrayList("i", "j")));
+    assertThat(fieldTypeNames, is(Lists.newArrayList("INTEGER", "INTEGER")));
```
Comment (danny0405) [inline core/src/test/java/org/apache/calcite/sql/type/SqlTypeUtilTest.java:156]: Try to use `Arrays.asList` instead which is from the JDK.

File: core/src/main/java/org/apache/calcite/sql/type/SqlTypeUtil.java [hunk 3]
```diff
@@ -1038,19 +1043,40 @@ public static SqlDataTypeSpec convertTypeToSpec(RelDataType type,
     // interval types, multiset types, etc
     assert typeName != null;
 
-    int precision = typeName.allowsPrec() ? type.getPrecision() : -1;
-    // fix up the precision.
-    if (maxPrecision > 0 && precision > maxPrecision) {
-      precision = maxPrecision;
+    SqlTypeNameSpec typeNameSpec;
+    if (isAtomic(type)) {
```
Comment (danny0405) [inline core/src/main/java/org/apache/calcite/sql/type/SqlTypeUtil.java:1047]: If possible, make `typeNameSpec ` final.

File: core/src/test/java/org/apache/calcite/sql/type/SqlTypeUtilTest.java (hunk 2 above)
Comment (Aaaaaaron) [reply core/src/test/java/org/apache/calcite/sql/type/SqlTypeUtilTest.java]: Thanks for your advice, I've changed my PR.

File: core/src/main/java/org/apache/calcite/sql/type/SqlTypeUtil.java (hunk 3 above)
Comment (Aaaaaaron) [reply core/src/main/java/org/apache/calcite/sql/type/SqlTypeUtil.java:1047]: Thanks for your advice, I've changed my PR.

Comment (danny0405) [reply core/src/main/java/org/apache/calcite/sql/type/SqlTypeUtil.java:1047]: Try `final SqlTypeNameSpec typeNameSpec` to see if it complains.

Comment (Aaaaaaron) [reply core/src/main/java/org/apache/calcite/sql/type/SqlTypeUtil.java:1047]: Thanks for your advice, I've changed my PR.

File: core/src/test/java/org/apache/calcite/sql/type/SqlTypeUtilTest.java [hunk 4]
```diff
@@ -117,6 +126,37 @@
...
+        .collect(Collectors.toList());
+    List<String> fieldTypeNames = rowSpec.getFieldTypes()
+        .stream()
+        .map(f -> f.getTypeName().getSimple())
+        .collect(Collectors.toList());
+    assertThat(rowSpec.getTypeName().getSimple(), is("ROW"));
+    assertThat(fieldNames, is(Arrays.asList("i", "j")));
+    assertThat(fieldTypeNames, is(Arrays.asList("INTEGER", "INTEGER")));
+    System.out.println();
+  }
```
Comment (danny0405) [inline core/src/test/java/org/apache/calcite/sql/type/SqlTypeUtilTest.java:158]: Remove `System.out.println();`.

Comment (Aaaaaaron) [reply core/src/test/java/org/apache/calcite/sql/type/SqlTypeUtilTest.java:158]: > Remove `System.out.println();`.
Ok

**Pull request closed without merging**